pkg/apis/serving/v1alpha1/revision_validation.go

```diff
@@ -389,3 +402,119 @@ func (current *Revision) CheckImmutableFields(og apis.Immutable) *apis.FieldErro
 
 	return nil
 }
+
+func validateNodeSelector(nodeSelector map[string]string) *apis.FieldError {
+	for key, value := range nodeSelector {
+		if errStrs :=
+			validation.IsQualifiedName(key); len(errStrs) > 0 {
+			return apis.ErrInvalidKeyName(
+				key,
+				key,
+				strings.Join(errStrs, "; "),
+			)
+		}
+		if errStrs :=
+			validation.IsValidLabelValue(value); len(errStrs) != 0 {
+			return &apis.FieldError{
+				Message: fmt.Sprintf("invalid value %q", value),
+				Details: strings.Join(errStrs, "; "),
+				Paths:   []string{key},
+			}
+		}
+	}
+	return nil
+}
+
+// validateToleration duplicates and adapts logic from
+// k8s.io/kubernetes/pkg/apis/core/validation. Although relevant functions from
+// that package are exported, they're not usable here because they are for
+// unversioned resources.
+func validateToleration(toleration corev1.Toleration) *apis.FieldError {
```

Comment:
Perma-linking for reference:

https://github.com/kubernetes/kubernetes/blob/ad4727440cc4a5be5c9714ffc0baefc11ce6c7f6/pkg/apis/core/validation/validation.go#L2834